sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -56,33 +57,107 @@ case class CreateArray(children: Seq[Expression]) extends Expression {
+  /**
+   * Return Java code pieces based on DataType and isPrimitive to allocate ArrayData class
+   *
+   * @param ctx a [[CodegenContext]]
+   * @param elementType data type of an underlying array
+   * @param elementsCode a set of [[ExprCode]] for each element of an underlying array
+   * @return (code pre-assignments, assignments to each array elements, code post-assignments,
+   *           arrayData name, underlying array name)
+   */
+  def genCodeToCreateArrayData(
+      ctx: CodegenContext,
+      elementType: DataType,
+      elementsCode: Seq[ExprCode],
+      allowNull: Boolean): (String, Seq[String], String, String, String) = {
+    val arrayName = ctx.freshName("array")
+    val arrayDataName = ctx.freshName("arrayData")
+    val numElements = elementsCode.length
+
+    if (!ctx.isPrimitiveType(elementType)) {
+      val arrayClass = classOf[ArrayData].getName
+      val genericArrayClass = classOf[GenericArrayData].getName
+      ctx.addMutableState("Object[]", arrayName,
+        s"this.$arrayName = new Object[${numElements}];")
+
+      val assignments = elementsCode.zipWithIndex.map { case (eval, i) =>
+        val isNullAssignment = if (allowNull) {
+          s"$arrayName[$i] = null;"
+        } else {
+          "throw new RuntimeException(\"Cannot use null as map key!\");"
+        }
+        eval.code + s"""
+         if (${eval.isNull}) {
+           $isNullAssignment
+         } else {
+           $arrayName[$i] = ${eval.value};
+         }
+       """
+      }
+
+      /*
+        TODO: When we declare arrayDataName as GenericArrayData,
+               we have to solve the following exception
+          https://github.com/apache/spark/pull/13909/files#r93813725
+      */
+      ("",
+       assignments,
+       s"final $arrayClass $arrayDataName = new $genericArrayClass($arrayName);",
+       arrayDataName,
+       arrayName)
+    } else {
+      val unsafeArrayClass = classOf[UnsafeArrayData].getName
+      val unsafeArraySizeInBytes =
+        UnsafeArrayData.calculateHeaderPortionInBytes(numElements) +
+        ByteArrayMethods.roundNumberOfBytesToNearestWord(elementType.defaultSize * numElements)
+      val baseOffset = Platform.BYTE_ARRAY_OFFSET
+      ctx.addMutableState(unsafeArrayClass, arrayDataName, "");
+
+      val primitiveValueTypeName = ctx.primitiveTypeName(elementType)
+      val assignments = elementsCode.zipWithIndex.map { case (eval, i) =>
+        val isNullAssignment = if (allowNull) {
+          s"$arrayDataName.setNullAt($i);"
+        } else {
+          "throw new RuntimeException(\"Cannot use null as map key!\");"
+        }
+        eval.code + s"""
+         if (${eval.isNull}) {
+           $isNullAssignment
+         } else {
+           $arrayDataName.set$primitiveValueTypeName($i, ${eval.value});
+         }
+       """
+      }
+
+      (s"""
+        byte[] $arrayName = new byte[$unsafeArraySizeInBytes];
+        $arrayDataName = new $unsafeArrayClass();
```

Comment:
As `arrayDataName` now is a mutable state and we point it to byte array for each iteration, we can just create the `UnsafeArrayData` once and reuse it. No need to create `UnsafeArrayData` each time.